includes/payments/class-edd-payment.php

```diff
@@ -1568,11 +1569,81 @@ private function process_refund() {
 
 		do_action( 'edd_pre_refund_payment', $this );
 
+		$decrease_store_earnings = apply_filters( 'edd_decrease_store_earnings_on_refund', true, $this );
+		$decrease_customer_value = apply_filters( 'edd_decrease_customer_value_on_refund', true, $this );
+		$decrease_purchase_count = apply_filters( 'edd_decrease_customer_purchase_count_on_refund', true, $this );
+
+		$this->maybe_alter_stats( $decrease_store_earnings, $decrease_customer_value, $decrease_purchase_count );
+		$this->delete_sales_logs();
+
+		// Clear the This Month earnings (this_monththis_month is NOT a typo)
+		delete_transient( md5( 'edd_earnings_this_monththis_month' ) );
+
+		do_action( 'edd_post_refund_payment', $this );
+	}
+
+	/**
+	 * Process when a payment is set to failed, decrement discount usages and other stats
+	 *
+	 * @since  2.5.7
+	 * @return void
+	 */
+	private function process_failure() {
+
+		$discounts = $this->discounts;
+		if ( 'none' === $discounts || empty( $discounts ) ) {
+			return;
+		}
+
+		if ( ! is_array( $discounts ) ) {
+			$discounts = array_map( 'trim', explode( ',', $discounts ) );
+		}
+
+		foreach ( $discounts as $discount ) {
+			edd_decrease_discount_usage( $discount );
+		}
+
+	}
+
+	private function process_pending() {
+		$process_pending = true;
+
+		// If the payment was not in publish or revoked status, don't decrement stats as they were never incremented
+		if ( ( 'publish' != $this->old_status && 'revoked' != $this->old_status ) || 'pending' != $this->status ) {
+			$process_pending = false;
+		}
+
+		// Allow extensions to filter for their own payment types, Example: Recurring Payments
+		$process_pending = apply_filters( 'edd_should_process_pending', $process_pending, $this );
+
+		if ( false === $process_pending ) {
+			return;
+		}
+
+		do_action( 'edd_pre_pending_payment', $this );
```

Comment:
I'm not sure this action name makes sense unless this method also fires when a pending payment is initially created.